build(docs-infra): switch the docs examples to Ivy and run unit tests for specific examples on CI

gkalpak opened the pull request:
This PR switches the docs examples to Ivy (from ViewEngine) and also sets up unit tests for specific docs examples to run on CI. See individual commits for more details.

This was originally part of #34374 (and split off for easier reviewing).
Partially addresses: [FW-1609](https://angular-team.atlassian.net/browse/FW-1609)

mary-poppins commented:
You can preview f998038 at https://pr36143-f998038.ngbuilds.io/.

petebacondarwin commented:
Oh and the 2nd commit has a typo: 

> is thruthy)

gkalpak commented on aio/content/examples/dependency-injection/example-config.json:
```diff
@@ -1,13 +0,0 @@
-      ]
-    }
-  ]
-}
```
This used a custom command, whose only difference from the default one was that it did not specify the `--prod` flag. This wasn't necessary any more (i.e. the tests pass with the `--prod` flag), so the custom command was no longer necessary.

gkalpak commented on aio/content/examples/http/example-config.json:
```diff
@@ -1,3 +1,7 @@
 {
-  "projectType": "testing"
+  "projectType": "testing",
+  "e2e": [
```
Yeah, thought of that too, but refrained from it (to avoid the extra noise). My hope is that in the future we will rename the whole script that runs these tests (which is currently called `run-example-e2e.js`) - and all the functions inside it (which mention e2e all the time) to the more generic `test`.

But happy to rename these properties now if you feel it is worth it.

gkalpak replied:
OK, I took the plunge and renamed the properties. I think it makes sense to do at least that for people looking at the config files (without knowning how the `run-example-e2e.js` script actually works).

PTAL

gkalpak commented on aio/content/examples/testing/src/app/welcome/welcome.component.spec.ts:
```diff
@@ -132,14 +132,4 @@ describe('WelcomeComponent', () => {
-    userServiceStub.isLoggedIn = false;
-    expect(userService.isLoggedIn).toBe(true);
-  });
-  // #enddocregion stub-not-injected
```
Exactly! (I thought I had left a comment on the PR, but apparently not.)

gkalpak commented on aio/content/examples/testing/src/tests.sb.ts:
```diff
@@ -18,6 +18,7 @@ import './app/hero/hero-detail.component.spec.ts';
 import './app/hero/hero-list.component.spec.ts';
 import './app/model/hero.service.spec.ts';
 import './app/model/testing/http-client.spec.ts';
+import './app/shared/canvas.component.spec.ts';
```
It seems so. (Also, this test would previously fail without the recent update to zone.js@0.10.3 :wink:)

gkalpak commented on aio/content/examples/upgrade-phonecat-1-typescript/tsconfig.json:
```diff
@@ -12,15 +12,15 @@
     "aot/**/*"
-  ],
-  "angularCompilerOptions": {
-    "enableIvy": false
```
Correct.

gkalpak commented on aio/content/examples/upgrade-phonecat-3-final/README.md:
```diff
@@ -3,9 +3,6 @@ structure.
 
-* Karma config for unit tests is in karma.conf.ng1.js because the boilerplate
-  Karma config is not compatible with the way tests in this project need to be run.
-  The shell script run-unit-tests.sh can be used to run the unit tests.
```
Yeah, the tests can be run via the normal karma config and there is no `run-unit-tests.sh` script in this example. (I assume this paragraph was copy-pasted from the previous two `upgrade-phonecat-*` examples (which do have the script).)

mary-poppins commented:
You can preview 6849e1b at https://pr36143-6849e1b.ngbuilds.io/.

mary-poppins commented:
You can preview 9689940 at https://pr36143-9689940.ngbuilds.io/.

mary-poppins commented:
You can preview 4876279 at https://pr36143-4876279.ngbuilds.io/.

gkalpak committed:
test(docs-infra): fix unit tests and run them for specific docs examples on CI

Previously, only e2e tests were run for docs examples on CI. As a
result, unit tests (which are included in the zipped archives we provide
for users to download and play with the examples locally) were often
outdated and broken.

This commit configures specific docs examples that have meaningful unit
tests to run them on CI (via the `run-example-e2e.js` script). Where
necessary, the unit tests are fixed to ensure they pass and reflect the
changes in the corresponding component/service.
This commit also removes some auto-generated unit tests that are not
meaningful (e.g. make trivial assertions, such that a component instance
is truthy) and are often broken anyway (e.g. because the corresponding
component has been changed in ways that make the tests fail).

gkalpak committed:
build(docs-infra): renamed `e2e` property of `example-config.json` to `tests`

Each docs example has an `example-config.json` configuration file. Among
other things, this file can be used to specify what commands to run in
order to test the example. (If not specified, the `run-example-e2e.js`
script will run a default `yarn e2e` command.)

Previously, the property specifying the test commands was called `e2e`.
This is because in the past only e2e tests were run for docs examples.
Since recently, some examples may specify commands for other types of
tests (such as unit tests). Therefore, calling the property that holds
the list of test commands `e2e` no longer makes sense and can be
misleading to people looking at the configuration files.

This commit renamed the property to the more generic `tests`. In the
future, the `run-example-e2e.js` script (and corresponding npm script)
should be renamed and refactored to also avoid giving the impression
that only e2e tests are run.

Discussed in:
https://github.com/angular/angular/pull/36143#discussion_r395148379

gkalpak committed:
build(docs-infra): update docs examples `package.json` templates wrt `core-js`

The `core-js` dependency is no longer included in `package.json` for
`cli`-type examples, but only for the `systemjs` ones. This commit
updates the `package.json` templates to reflect that (and also updates
the `npm-packages` guide accordingly).

mary-poppins commented:
You can preview b5f3156 at https://pr36143-b5f3156.ngbuilds.io/.

kara reviewed (approved):
LGTM

Reviewed-for: global-approvers

kara commented:
Reviewed-for: global-approvers

kara commented:
@gkalpak The merge script is running into merge conflicts on master for some reason. Can you rebase to see if that helps?

gkalpak committed:
build(docs-infra): switch docs examples to Ivy

The docs examples are switched to Ivy. On CI, the tests are run with
both Ivy and ViewEngine.

Partially addresses:
[FW-1609](https://angular-team.atlassian.net/browse/FW-1609)
